tommath.h

```diff
@@ -430,7 +430,7 @@ mp_err mp_lcm(const mp_int *a, const mp_int *b, mp_int *c) MP_WUR;
  *
  * returns error if a < 0 and b is even
  */
-mp_err mp_root_u32(const mp_int *a, uint32_t b, mp_int *c) MP_WUR;
+mp_err mp_root(const mp_int *a, int b, mp_int *c) MP_WUR;
```

Comment:
Yes, the convention in ltm seems to be to use _d even if it has nothing to do with digits.
I remember that we had this discussion, when we did the renaming. Back then I argued against the suffix. But I am actually ok with a suffix, the issue is just that _u32 is overspecific. This is not what we want. In particular since the function might not even work for all uint32_ts, that is the problem. At least I found many fishy casts inside those functions...

I think the only functions which should be suffixed with the very specific suffix are the init and set functions, since they are very useful for example for language bindings.

Reply:
The suffix `_d` has been implemented to differentiate between two function that do the same but one accepts a native data type instead of a bigint. I think it makes sense to keep such a suffix.

Then there are the weirder names like .e.g.: `mp_mul_2d` for shifting left bitwise. It means multiplying with `2^x`, so the `mp_mul` part is OK and the `_2` part, too, but the final `_d` is only there because of `mp_mul_2` (which I think should be a private function and called by `mp_mul_2d`).
Then there is `mp_lshd`, which does shifting left limb-wise. That is multiplying with `n *2^{beta}` but no `mp_mul` part here.(This could also go into the private room and called from `mp_mul_2d` instead and all together renamed to something like e.g.: `mp_shift_left`).

The two functions with the suffix `_u32` don't have counterparts, there is no reason to make a difference, it does not need a suffix which already got removed while I'm writing this, I'm too slow, sorry.

Since MP_8BIT is gone we can safely assume to have one native unsigned integral data type with at least 32 bit capacity and we should use it as the input where it was `mp_digit` or `int` before to give the input a well defined size.

But, as @minad already noted, "[T]he function might not even work for all `uint32_t`", that needs to get repaired first. If it is possible. Is it? It is no problem in the cases of `mp_expt_32` and `mp_log_32`. It will get a bit more complicated with e.g.: `mp_mul_d` but not that much.

Then there are the cases where the input needs to be signed and which has `int` now. One example is `mp_prime_is_prime` where it is used to transport a value and where the sign acts as a flag. This was due to legacy reasons and can get changed now, I think.

The rest is either a size and will get changed to `size_t` if I understood the discussion here correctly, or is a simple true/false flag, e.g.: `redmode` which should be changed to `bool` instead. (The `int` in `mp_kronecker` needs three states `{-1, 0, 1}`).

Oh, and thanks for getting rid of the unnecessary pointer juggling and replacing it with direct access to the limb-array!

Reply:
@czurnieden 

> The two functions with the suffix _u32 don't have counterparts, there is no reason to make a difference, it does not need a suffix which already got removed while I'm writing this, I'm too slow, sorry.

This is what I argued a long time ago, but @sjaeckel prefers a suffix. What does _d mean? It means digit right or we say it means "native **d**atatype" and use it for those functions too?

> Oh, and thanks for getting rid of the unnecessary pointer juggling and replacing it with direct access to the limb-array!

I would really appreciate if you can look particularly at those accesses while reviewing. I hope I didn't mess anything up while doing it. I am mostly interested in this simplifications since this will enable me to implement full width digits in the end.

> The rest is either a size and will get changed to size_t if I understood the discussion here correctly

Regarding size_t, see what I wrote in #430. I am still thinking about how to proceed.

Reply:
> Using no suffix or use the weird _d or something else?

No suffix for `mp_log` and `mp_expt`, as they have no bigint counterpart.

Suffix `_d` for functions that have a bigint counterpart like e.g.: `mp_mul_d` no matter what the size (and signedness) of the small input data type. It is as good as anything else, wouldn't change it without a good reason.

For the rest like e.g.: `mp_mul_2d`? Good question. I would leave it for now and see how the rest develops. Renaming functions after a couple of decades is always frowned upon by the old users but sometimes it is just necessary.

> I am mostly interested in this simplifications since this will enable me to implement full width digits in the end.

Using all bits would <del>ruin</del> slow down my FFT implementation by quite some amount. Not a problem for LTM, I know but what is with the Comba multiplication?
How much will it slow down the assembler-free C-only version?
Questions upon questions! ;-)

> Regarding size_t, see what I wrote in #430. I am still thinking about how to proceed.

My bet is on `size_t` ;-)

From #435 
> [&hellip;] and simplifying mp_prime_strong_lucas_selfridge, s_mp_exptmod, s_mp_exptmod_fast 

The function `mp_prime_strong_lucas_selfridge` is a verbatim port and security related and I don't want to touch it without an urgent need. What exactly are your problems with it (is it for the "use all of the bits of the limbs"?)?

But I will take a look at the other two, of course.

Reply:
suffix: I would also not use a suffix, but @sjaeckel needs convincing. The question is if we will add full mp_int versions at some point where all arguments are mp_int.

I have no serious problems with the function mp_prime_strong_lucas_selfridge, s_mp_exptmod, s_mp_exptmod_fast. They look just very complex. In particular if you are saying that they are security related, complexity is evil ;)

comba: I have no idea if it will be significantly slower than the current comba, since I have not done it. Maybe it will be bad. But I am thinking about abstracting **add with carry** and **wide multiplication** somehow. Then we can compile the library in multiple modi, maybe even like it is now. For comba I have to introduce a accumulator type then. 

Regarding FFT: It is sad that it will be slower but I think FFT multiplication is not relevant for most applications of ltm (e.g. language runtimes), but it is still nice to have.